File: python_api/SparkVersion/Operations/Helpers/StatisticsCreator.py
from pyspark.sql.functions import col
from pyspark.sql.types import StringType
from pyspark.sql import functions as F

from Models.DataStatistics import DataStatistics
from Models.IntStringTuple import IntStringTuple
from Models.NumberColumnsSummary import NumberColumnsSummary
from Models.TextColumnsSummary import TextColumnsSummary


def createStatisticsSummary(data):
    return DataStatistics(textColumnsSummary(data), numberColumnsSummary(data))


def textColumnsSummary(data):
    textColumnNames = [field.name for field in data.schema.fields if (isinstance(field.dataType, StringType))]
    textColumn = data.select(*textColumnNames)
    textColumnsSummaries = [textColumnsSummaryFromNames(name, textColumn) for name in textColumnNames]
    return textColumnsSummaries


def numberColumnsSummary(data):
    numberColumnNames = [field.name for field in data.schema.fields if not (isinstance(field.dataType, StringType))]
    numberColumn = data.select(*numberColumnNames)
    summary = numberColumn.summary()
    numberColumnsSummaries = [numberColumnsSummaryFromNames(name, summary) for name in numberColumnNames]
    return numberColumnsSummaries


def textColumnsSummaryFromNames(name, textColumn):
    count = textColumn.count()
    df_temp = textColumn.select(name) \
        .groupBy(name) \
        .agg((F.count(F.lit(1)) / textColumn.count()).alias("frequency")) \
        .sort(col("frequency").desc())
    distinct = df_temp.count()
    bestName = df_temp.collect()[0].__getitem__(name)
    bestCount = df_temp.collect()[0].__getitem__("frequency")
    secondName = df_temp.collect()[1].__getitem__(name)
    secondCount = df_temp.collect()[1].__getitem__("frequency")
    mostCommons = [
        IntStringTuple(bestName, bestCount),
        IntStringTuple(secondName, secondCount)
    ]
    summary = TextColumnsSummary(name, count, distinct, mostCommons)
    return summary


def numberColumnsSummaryFromNames(name, summary):
    return NumberColumnsSummary(
        name = name,
        count = summary.collect()[0].__getitem__(name),
        mean = summary.collect()[1].__getitem__(name),
        stdDev = summary.collect()[2].__getitem__(name),
        min = summary.collect()[3].__getitem__(name),
        percentage25 = summary.collect()[4].__getitem__(name),
        percentage50 = summary.collect()[5].__getitem__(name),
        percentage75 = summary.collect()[6].__getitem__(name),
        max = summary.collect()[7].__getitem__(name),
    )

File: scala_api/src/main/scala/application/dataAnalysis/operations/Helpers/StatisticsCreator.scala
package application.dataAnalysis.operations.Helpers

import application.models.statistics.{DataStatistics, IntStringTuple, NumberColumnsSummary, TextColumnsSummary}
import org.apache.spark.sql.{DataFrame, Row}
import org.apache.spark.sql.functions.col
import org.apache.spark.sql.types.{StringType, TimestampType}

class StatisticsCreator {

  def createStatisticsSummary(data: DataFrame): DataStatistics = {
    DataStatistics(textColumnsSummary(data), numberColumnsSummary(data))
  }

  private def textColumnsSummary(data: DataFrame): Seq[TextColumnsSummary] = {
    val textColumnNames: Seq[String] = data.schema.filter(
      x => x.dataType == StringType || x.dataType == TimestampType
    ).map(_.name)
    val textColumn = data.select(textColumnNames.map(col): _*)
    textColumnNames.map(name => {
      val rdd = textColumn.select(name).rdd
      val distinct = rdd.distinct().count()
      val allCount = rdd.count()
      val mostCommons = rdd.map(s => (s, 1))
        .reduceByKey(_ + _)
        .map { case (s, count) => (count, s) }
        .top(3)(Ordering.by(_._1))
        .map { case (count, s) => IntStringTuple(count.toDouble / allCount, s.get(0).toString)}
      TextColumnsSummary(name, allCount, distinct, mostCommons.toSet)
    })
  }

  private def numberColumnsSummary(data: DataFrame): Seq[NumberColumnsSummary] = {
    val numberColumnNames: Seq[String] = data.schema.filter(x => x.dataType != StringType).map(_.name)
    val numberColumn = data.select(numberColumnNames.map(col): _*)
    val summary = numberColumn.summary()
    numberColumnNames.map { name =>
      val rdd: Array[Row] = summary.select(col(name)).rdd.take(8)
      NumberColumnsSummary(
        name = name,
        count = rdd(0).get(0).toString.toDouble.toLong,
        mean = rdd(1).get(0).toString.toDouble,
        stdDev = rdd(2).get(0).toString.toDouble,
        min = rdd(3).get(0).toString.toDouble,
        percentage25 = rdd(4).get(0).toString.toDouble,
        percentage50 = rdd(5).get(0).toString.toDouble,
        percentage75 = rdd(6).get(0).toString.toDouble,
        max = rdd(7).get(0).toString.toDouble
      )
    }
  }
}
